File: apps/ios/TypviaKit/Sources/Screens/RoomBar.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import SwiftUI

/// The four rooms, as words.
///
/// No icons, no tab bar chrome, no pill behind the current room: the room you
/// are in is the one with a caret in front of it. The bar sits on the sheet
/// with nothing under it, because the delivery's list of forbidden things
/// includes a bottom bar that covers content.
public struct RoomBar: View {
    /// The rooms that have a door. `ai` is not among them: it is a capability
    /// the other rooms use, not a place.
    public static let rooms: [Room] = [.home, .library, .vault, .settings]

    @Environment(\.tr) private var tr

    private let current: Room
    private let available: Set<Room>
    private let onInkRoom: Bool
    private let select: ((Room) -> Void)?

    /// - Parameter available: the rooms that have been built. The others are
    ///   still printed — the bar is the product's map and hiding a room would
    ///   misdraw it — but they are text rather than controls, because a word
    ///   that does nothing when tapped is worse than a word that is plainly
    ///   not a door yet.
    /// - Parameter onInkRoom: the bar sits on a dark room — the shut chapter,
    ///   and the vault when it arrives — and takes that room's colours.
    public init(
        current: Room,
        available: Set<Room> = [],
        onInkRoom: Bool = false,
        select: ((Room) -> Void)? = nil
    ) {
        self.current = current
        self.available = available
        self.onInkRoom = onInkRoom
        self.select = select
    }

    public var body: some View {
        HStack(spacing: 0) {
            ForEach(RoomBar.rooms, id: \.self) { room in
                entry(for: room)
            }
        }
        .padding(.top, Metrics.topPadding)
        .padding(.bottom, Metrics.bottomPadding)
        // The bar is a strip across the foot of the screen, not a row of words
        // pushed to one side. A hairline separates it from the page so it does
        // not read as more content.
        .frame(maxWidth: .infinity)
        .overlay(alignment: .top) { Hairline() }
    }

    @ViewBuilder
    private func entry(for room: Room) -> some View {
        let isCurrent = room == current
        let label = VStack(spacing: Metrics.iconGap) {
            RoomIcon(room: room, isCurrent: isCurrent)
                .foregroundStyle(ink(isCurrent: isCurrent))
            Text(name(of: room))
                .typviaType(.monoLabel)
                .foregroundStyle(ink(isCurrent: isCurrent))
        }
        // Each door takes an equal quarter and the platform's full minimum
        // height, so the whole column is the target rather than the word.
        .frame(maxWidth: .infinity, minHeight: Tokens.Hit.minimum)
        .contentShape(Rectangle())
        .accessibilityElement(children: .combine)
        .accessibilityLabel(name(of: room))
        .accessibilityAddTraits(isCurrent ? [.isSelected] : [])

        if let select, !isCurrent, available.contains(room) {
            Button { select(room) } label: { label }
                .buttonStyle(.plain)
        } else {
            label
        }
    }

    /// Which door you are standing in is said twice — the icon and its word
    /// both take full ink, the other three sit at meta. Colour alone never
    /// carries a meaning in this product, and here it does not have to: the
    /// weight of the type differs as well.
    private func ink(isCurrent: Bool) -> Color {
        if isCurrent {
            return onInkRoom ? VaultRoom.ink : Paper.ink
        }
        return onInkRoom ? VaultRoom.ink3 : Paper.ink3
    }

    private func name(of room: Room) -> String {
        switch room {
        case .home: tr("Home", "首页")
        case .library: tr("Library", "资料库")
        case .vault: tr("Vault", "保险库")
        case .settings: tr("Settings", "设置")
        case .ai: tr("AI", "AI")
        }
    }

    /// Taken from the frames, which draw this bar identically on every screen.
    private enum Metrics {
        static let iconGap: CGFloat = 5
        static let topPadding: CGFloat = 10
        static let bottomPadding: CGFloat = 4
    }
}

File: apps/android/core-ui/src/main/kotlin/dev/typvia/mobile/ui/RoomBar.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package dev.typvia.mobile.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

/**
 * The four doors, as words.
 *
 * No icons: the navigation is set in the same type as everything else, and
 * where the reader is standing is said by a caret beside the word rather than
 * by a colour or a filled shape. A room with no screen behind it is not
 * printed — a word in this bar is never a dead end.
 */
@Composable
fun RoomBar(
    current: Room,
    available: Set<Room>,
    modifier: Modifier = Modifier,
    onInkRoom: Boolean = false,
    select: ((Room) -> Unit)? = null,
) {
    val tr = LocalTranslator.current
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(if (onInkRoom) VaultRoom.base else Paper.base),
    ) {
        // A rule holds the bar off the page, so it does not read as more
        // content.
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(Tokens.Line.hairlineWidth)
                .background(Paper.rule),
        )
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    top = RoomBarMetrics.top,
                    // Gesture navigation leaves a strip at the bottom, and the
                    // delivery keeps the paper above it rather than under it.
                    bottom = Tokens.Viewport.bottomSafeArea,
                ),
        ) {
            for (room in RoomBarOrder.rooms) {
                val isHere = room == current
                val isOpen = room in available
                val ink = when {
                    onInkRoom && isHere -> VaultRoom.ink
                    onInkRoom -> VaultRoom.ink3
                    isHere -> Paper.ink
                    else -> Paper.ink3
                }
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(RoomBarMetrics.iconGap),
                    modifier = Modifier
                        // An equal quarter each, and the whole column is the
                        // target rather than the word on it.
                        .weight(1f)
                        .heightIn(min = Tokens.Hit.minimum)
                        .clickable(enabled = select != null && !isHere && isOpen) {
                            select?.invoke(room)
                        }
                        .semantics(mergeDescendants = true) { selected = isHere },
                ) {
                    RoomIcon(room = room, tint = ink)
                    Text(
                        text = room.title(tr),
                        style = TypviaType.MonoLabel.style(tr.language),
                        color = ink,
                        textAlign = TextAlign.Center,
                    )
                }
            }
        }
    }
}


/** The order the doors are printed in — the reading order of the product. */
object RoomBarOrder {
    val rooms = listOf(Room.Home, Room.Library, Room.Vault, Room.Settings)
}

fun Room.title(tr: Translator): String = when (this) {
    Room.Home -> tr("Home", "首页")
    Room.Library -> tr("Library", "资料库")
    Room.Vault -> tr("Vault", "保险库")
    Room.Settings -> tr("Settings", "设置")
    Room.Ai -> tr("AI", "AI")
}

object RoomBarMetrics {
    /** Above the row of doors; the strip's own breathing room. */
    val top = 10.dp

    /** Between a door's mark and its word. */
    val iconGap = 4.dp

}
